packages/flame/CHANGELOG.md

```diff
@@ -46,6 +46,11 @@
  - Added `FocusNode` on the game widget and improved keyboard handling in the game.
  - Added ability to have custom mouse cursor on the `GameWidget` region
  - Change sprite component to default to the Sprite size if not provided
+ - `Game` is now a `Component`
+ - `ComponentEffect` is now a `Component`
+ - `HasGameRef` can now operate independently from `Game`
+ - `preOffset` and `postOffset` for effects to handle time before and after an effect
+ - Removes `component.onMount` since `onLoad` can be used for all its purposes
```

Comment:
The current implementation allows putting expensive initialization code into the `onLoad` method, which serves as a late-stage constructor, and is guaranteed to run only once. The `onMount`, on the other hand, runs every time the component is mounted into the game tree. For example, if the component is removed from the game, and then attached at a new place in the tree, we'd run `onRemove` and then `onMount` again. Thus, `onMount` can be used to perform initialization that requires awareness of the surrounding game tree.

So, how are these tasks going to be separated now?

Reply:
`onLoad` would also run again if it changed parent, since that is part of `ComponentSet.addChild`.
Maybe it shouldn't though.

Reply:
This could potentially be problematic if `onLoad` initializes `late final` variables (which it usually does).

Reply:
Good point, maybe there should be a new lifecycle method called `onParentChange` or something like that which runs after `onLoad` (`onMount` ran before).

Reply:
I just tested this with a couple of print statements, and in the current `main` branch the sequence of events is `onGameResize -> onLoad -> onMount -> (update -> render)*`.

Reply:
Alright, then it was just wrong in the docs. What do you think of the new name?

Reply:
Frankly, when comparing `onParentChange` vs `onMount`, I like the latter slightly more. Maybe because I'm already used to it, but also because it pairs better with `onRemove`.

So, am I correct in understanding that `onLoad` will now run only when the component is already mounted? I mean, this is obviously nice to have if you need to access `gameRef` within the `onLoad` handler, but adds a bit more confusion with respect to the fact that now as the component is mounted we run `onLoad` first, before the `onMount`. I guess it should be explained really clearly in the docs then.

Reply:
I think it sounds better with `onMount`, but I think it is really vague what it actually means. Especially when it is changing parents, is it mounted again on that tree? `onMount` sounds more like a synonym to `onAttach` to me.

I have updated the section in the docs about the lifecycle of the `Component` and the `Game` a bit.

This is how the lifecycle looks now:
https://github.com/flame-engine/flame/blob/ad385c16713ffc44e2d64510bac7a13ea5be74db/doc/images/component_lifecycle.png

Reply:
But `onAttach` is for the `Game` class only, not for a generic component, right? It could have a more verbose name in that case, something like `onFlutterTreeAttach`.

As for the vagueness of `onMount`, I guess it would help if we started by clarifying what the `isMounted` property actually means. We could start with the notion of the component tree, and how it can be traversed up and down. That component is called "mounted" when it is placed inside that tree. That there are mechanisms in place to ensure that all components in the tree will get updated/rendered on every game tick, and that they will properly receive input events such as taps or mouse clicks.

Then we would say that a component can be removed (detached) from the component tree, in which case it will be considered "unmounted", and then added back into the tree, maybe at another place, in which case call the `onMount` handler again.

As for `onParentChange` -- it immediately raises several questions: (1) is `onRemove` called at all when you change a component's parent? (2) if you remove the component from a tree and then later add it back for the same parent -- does this qualify as a "parent change"?

Reply:
True, that could be a good name for `onAttach`/`onDetach`.

Doesn't `onMount` raise the same questions as the ones that you spell out in the last paragraph?

Reply:
The second question is avoided because "onMount" doesn't imply a change of parent -- only that the component is being put into the component tree. So I know that every time a component is being placed into a tree, the game calls `onMount`. The first question becomes trivial if we state that for every `onMount` there will be a call to `onRemove` later on. So that even though a parent change can be as simple as `this.parent = newParent`, we would still do it in two steps.

Reply:
I don't know, I still feel that `onMount` needs more explanation than `onParentChange` by just reading their names (calling `changeParent` with the same parent really is an edge case, which will be explained in the docs).

Reply:
I would say, a thorough explanation of when the method is called is needed in any case. Then it is the matter of having a name that best matches that explanation.

There is a flag `isMounted` to describe whether the component is currently in the game tree or not. This flag is more harmonious with the `onMount` name. Or are you planning to change the flag's name as well?

To me, `onParentChange` sounds like it is called whenever the component's `parent` property changes. Which leaves one wondering whether it'll be called (a) only when the `parent` property changes, or also when it is initialized? (b) when `parent` property becomes `null` because the component is removed from the tree.

Reply:
I talked to the team about what they thought and they got quite confused about `onParentChange` so I will rename it to `onMount` and re-add `isMounted` (since that was removed, but I don't remember what my reasoning for removing it was).
I'll try to make it clear how it works in the docs too. :)